src/server/game/Anticheat/AnticheatMgr.h

```diff
@@ -0,0 +1,123 @@
+/*
+* This file is part of Project SkyFire https://www.projectskyfire.org.
+* See LICENSE.md file for Copyright information
+*/
+
+#ifndef SC_ACMGR_H
+#define SC_ACMGR_H
+
+#include "Common.h"
+#include "SharedDefines.h"
+#include <unordered_map>
+#include "WorldSession.h"
+
+class Player;
+class ChatHandler;
+
+enum ReportTypes
+{
+    SPEED_HACK_REPORT = 0,
+    FLY_HACK_REPORT = 1,
+    WALK_WATER_HACK_REPORT = 2,
+    JUMP_HACK_REPORT = 3,
+    TELEPORT_PLANE_HACK_REPORT = 4,
+    CLIMB_HACK_REPORT = 5,
+    TIME_MANIPULATION_REPORT = 6,
+    GRAVITY_HACK_REPORT = 7,
+    TELEPORT_HACK_REPORT = 8,
+    BG_START_HACK_REPORT = 9,
+    MAX_REPORT_TYPES
+};
+
+struct AnticheatData
+{
+    AnticheatData();
+
+    void SetPosition(float x, float y, float z, float o);
+    void Reset();
+
+    uint32 lastOpcode;
+    MovementInfo lastMovementInfo;
+    uint32 totalReports;
+    float average;
+    uint32 creationTime;
+    bool hasDailyReport;
+
+    uint32 GetTotalReports() const;
+    void SetTotalReports(uint32 _totalReports);
+
+    uint32 GetTypeReports(uint32 type) const;
+    void SetTypeReports(uint32 type, uint32 amount);
+
+    float GetAverage() const;
+    void SetAverage(float _average);
+
+    uint32 GetCreationTime() const;
+    void SetCreationTime(uint32 creationTime);
+
+    uint32 typeReports[MAX_REPORT_TYPES];
+    uint32 tempReports[MAX_REPORT_TYPES];
+    uint32 tempReportsTimer[MAX_REPORT_TYPES];
+
+    uint32 reportAnnounceCount[MAX_REPORT_TYPES];
+    time_t reportAnnounceCooldown[MAX_REPORT_TYPES];
+    void SetDailyReportState(bool b);
+    bool GetDailyReportState();
+};
+
+// GUIDLow is the key.
+typedef std::unordered_map<uint32, AnticheatData> AnticheatPlayersDataMap;
+
+class AnticheatMgr
+{
+    AnticheatMgr();
+    ~AnticheatMgr();
+
+    public:
+        static AnticheatMgr *instance()
```

Comment:
friend class ACE_Singleton<ObjectMgr, ACE_Null_Mutex>;

Reply:
thank you.
48a1d9749a